bot/cogs/moderation/infractions.py

```diff
@@ -0,0 +1,519 @@
+import logging
+import textwrap
+import typing as t
+
+import dateutil.parser
+import discord
+from discord import Member
+from discord.ext import commands
+from discord.ext.commands import Context, command
+
+from bot import constants
+from bot.api import ResponseCodeError
+from bot.constants import Colours, Event
+from bot.converters import Duration
+from bot.decorators import respect_role_hierarchy
+from bot.utils import time
+from bot.utils.checks import with_role_check
+from bot.utils.scheduling import Scheduler
+from . import utils
+from .modlog import ModLog
+from .utils import MemberObject
+
+log = logging.getLogger(__name__)
+
+MemberConverter = t.Union[utils.UserTypes, utils.proxy_user]
+
+
+class Infractions(Scheduler, commands.Cog):
+    """Server moderation tools."""
+
+    def __init__(self, bot: commands.Bot):
+        self.bot = bot
+        self.category = "Moderation"
+        self._muted_role = discord.Object(constants.Roles.muted)
+        super().__init__()
+
+    @property
+    def mod_log(self) -> ModLog:
+        """Get currently loaded ModLog cog instance."""
+        return self.bot.get_cog("ModLog")
+
+    @commands.Cog.listener()
+    async def on_ready(self) -> None:
+        """Schedule expiration for previous infractions."""
+        infractions = await self.bot.api_client.get(
+            'bot/infractions',
+            params={'active': 'true'}
+        )
+        for infraction in infractions:
+            if infraction["expires_at"] is not None:
+                self.schedule_task(self.bot.loop, infraction["id"], infraction)
+
+    # region: Permanent infractions
+
+    @command()
+    async def warn(self, ctx: Context, user: MemberConverter, *, reason: str = None) -> None:
+        """Warn a user for the given reason."""
+        infraction = await utils.post_infraction(ctx, user, reason, "warning")
```

Comment:
The reason and the warning type were switched here
```suggestion
        infraction = await utils.post_infraction(ctx, user,  "warning", reason)
```

`!warn @David Bowie Kitties!` led to:
```py
 Oct 02 21:22:29 Bot: |         bot.cogs.moderation.utils |    ERROR | An unexpected ResponseCodeError occurred while adding an infraction:
Traceback (most recent call last):
  File "/home/sebastiaan/pydis/repositories/djangobot/bot/cogs/moderation/utils.py", line 71, in post_infraction
    response = await ctx.bot.api_client.post('bot/infractions', json=payload)
  File "/home/sebastiaan/pydis/repositories/djangobot/bot/api.py", line 76, in post
    await self.maybe_raise_for_status(resp, raise_for_status)
  File "/home/sebastiaan/pydis/repositories/djangobot/bot/api.py", line 56, in maybe_raise_for_status
    raise ResponseCodeError(response=response, response_json=response_json)
bot.api.ResponseCodeError: Status: 400 Response: {'type': ['"Kitties!" is not a valid choice.']}
```